math/arcsine.go

```diff
@@ -1,69 +1,50 @@
-
-2. Redistributions in binary form must reproduce the above copyright notice,
-	this list of conditions and the following disclaimer in the documentation and/or
-	other materials provided with the distribution.
-
-3. Neither the name of the copyright holder nor the names of its contributors may be
-	used to endorse or promote products derived from this software without specific
-	prior written permission.
-
-THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
-EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
-OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
-SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
-SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
-SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
-HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
-OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
-SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-*/
 import (
 	"github.com/ericlagergren/decimal"
+	"github.com/ericlagergren/decimal/misc"
 )
 
-//Asin returns the arcsine value in radians.
-// Input range : -1 <= value <= 1
-// Output range: -pi/2 <= Asin() <= pi/2
-// Notes:
-//		Asin(NaN)       -> NaN
-//		Asin(+/-Inf)    -> NaN
-//		Asin(|value|>1) -> NaN
-//		Asin(+/-1)      -> +/-pi/2
-func Asin(z *decimal.Big, value *decimal.Big) *decimal.Big {
-	// here we'll use the formula
-	// Asin(x) = 2atan(x/(1+sqrt(1-x*x)))
-	calculatingPrecision := z.Context.Precision + defaultExtraPrecision
+// Asin returns the arcsine, in radians, of x.
+//
+// Range:
+//     Input: -1 <= x <= 1
+//     Output: -pi/2 <= Asin(x) <= pi/2
+//
+// Special cases:
+//		Asin(NaN)  = NaN
+//		Asin(±Inf) = NaN
+//		Asin(x)    = NaN if x < -1 or x > 1
+//		Asin(±1)   = ±pi/2
+func Asin(z, x *decimal.Big) *decimal.Big {
+	if z.CheckNaNs(x, nil) {
+		return z
+	}
 
-	if value.IsInf(0) || value.IsNaN(0) || one.CmpAbs(value) < 0 {
+	cmp1 := x.CmpAbs(one)
+	if x.IsInf(0) || cmp1 > 0 {
 		z.Context.Conditions |= decimal.InvalidOperation
 		return z.SetNaN(false)
 	}
 
-	if one.CmpAbs(value) == 0 {
-		piOver2 := Pi(z)
-		piOver2.Quo(piOver2, two)
+	ctx := decimal.Context{Precision: precision(z)}
 
-		if value.Signbit() {
-			piOver2.Neg(piOver2)
+	if cmp1 == 0 {
+		pi2(z, ctx)
+		if x.Signbit() {
+			misc.SetSignbit(z, true)
 		}
-
-		return z.Set(piOver2)
+		return z
 	}
 
-	xsq := decimal.WithPrecision(calculatingPrecision).Mul(value, value)
-	x := xsq.Quo(value, xsq.Add(Sqrt(xsq, xsq.Sub(one, xsq)), one))
-	result := Atan(decimal.WithPrecision(calculatingPrecision), x)
+	ctx.Precision += defaultExtraPrecision
+
+	// Asin(x) = 2 * atan(x / (1 + sqrt(1 - x*x)))
 
-	result = result.Mul(two, result)
-	return z.Set(result)
+	x2 := ctx.Mul(alias(z, x), x, x)
```

Comment:
I'm not sure I understand the purpose of alias.  If ctx.Mul() doesn't check if z is aliased why should this function? Isn't that just part of the agreement that z will not be x?

Reply:
We still need to use x afterwards, so the call to alias makes sure that we only use z if it != x, that way we don’t overwrite x.